### PR TITLE
feat(list-view): add basic list view with and without expansion

### LINKED ISSUE #30: Component - List
Implement a component that provides the following UI element:

- [PatternFly List View](http://www.patternfly.org/pattern-library/content-views/list-view/)

Adds the new List styles here:
https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-pf.html

Started in #63 . Needs help ;)

Acceptance Criteria:

- When rendered, the component matches the examples in the PatternFly Pattern Library (see link above)

### PR DESCRIPTION
**Pull request by waldenraines**
Adds a first pass at expandable and non-expandable list view based on patternfly's markup.

Closes #30

**What**:

Creates the patternfly list view layout in both [simple expansion](http://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-simple-expansion.html) and [non expandable](http://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view.html) forms.  There are still some features to be added (such as the action buttons, the select checkbox, etc.) but this is a good first pass.

**Why**:

I needed to use both of these list view layouts in my application and they are not yet implemented in patternfly-react.

**How**:

See below.  This was based on https://github.com/patternfly/patternfly-react/pull/31 and contains all of the additional changes listed there.  Ideally I think we should make those additional changes in a separate PR and if that happens I will rebase this.

@priley86 sorry to create a somewhat of a competing PR.  I believe this component here should be called `<ListView>` and your component from https://github.com/patternfly/patternfly-react/pull/31 should maybe be called `<List>` or `<ListPf>` or something in accordance with how items are named [here](http://rawgit.com/patternfly/patternfly/master-dist/dist/tests/).

Thank you to @matanwerbner for his assistance with this PR.

Storybook is [here](https://waldenraines.github.io/patternfly-react/).

**THIS PR REQUIRES https://github.com/patternfly/patternfly-react/pull/35**

### CODE REVIEW

File: src/ListView/__snapshots__/ListViewExpandableItem.test.js.snap [hunk 1]
```diff
@@ -0,0 +1,194 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (ohadlevy) [inline src/ListView/__snapshots__/ListViewExpandableItem.test.js.snap]: should not be part of the pr ?

**Comment by ohadlevy** [comment]
@waldenraines also note test failures.

File: src/ListView/__snapshots__/ListViewExpandableItem.test.js.snap (hunk 1 above)
Comment (waldenraines) [inline src/ListView/__snapshots__/ListViewExpandableItem.test.js.snap]: It is supposed to be actually

**Comment by waldenraines** [comment]
> thanks @waldenraines ! any chance you can push our stories to some github page ?

Oh thanks!  In the process of filling out the PR form I forgot to add that.  Maybe we should make that one of the questions you have to answer when creating a PR.

I added the storybook to the PR description above.

File: src/ListView/__snapshots__/ListViewExpandableItem.test.js.snap (hunk 1 above)
Comment (ohadlevy) [inline src/ListView/__snapshots__/ListViewExpandableItem.test.js.snap]: oh - thanks I didn't know about it :)

**Comment by ohadlevy** [comment]
@waldenraines it makes sense, one thing that is needs to be remembered that you need a per branch storybook otherwise if you have multiple pr's you will get the storybook of only one of them at a time.

**Comment by waldenraines** [comment]
> @waldenraines also note test failures.

I'm currently looking into this.  My hunch is that it has to do with a difference in version of [prettier](https://github.com/prettier/prettier) that I have locally versus what is on the travis.  For example those alert tests fail for me in master as well, see also https://github.com/patternfly/patternfly-react/pull/34

**Comment by waldenraines** [comment]
> @waldenraines it makes sense, one thing that is needs to be remembered that you need a per branch storybook otherwise if you have multiple pr's you will get the storybook of only one of them at a time.

patternfly-react uses [storybook-deployer](https://github.com/storybooks/storybook-deployer) via `npm run storybook:deploy` to create the github pages storybook. 

It would be nice if storybook-deployer's `storybook_to_ghpages` automatically supported this instead of the branch being [hardcoded to master](https://github.com/storybooks/storybook-deployer/blob/master/bin/storybook_to_ghpages#L56).

File: src/ListView/__snapshots__/ListViewExpandableItem.test.js.snap (hunk 1 above)
Comment (waldenraines) [inline src/ListView/__snapshots__/ListViewExpandableItem.test.js.snap]: It's how they track the version of jest the snapshot was created with I guess?

File: src/ListView/ListViewItemContents.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  heading: PropTypes.node.isRequired,
+  /** item text node  */
+  itemText: PropTypes.node,
+  /** additional info node  */
+  additionalInfo: PropTypes.node
+}
+
+ListViewItemContents.defaultProps = {
+  /** default icon class */
+  iconClass: 'fa-question',
```
Comment (waldenraines) [inline src/ListView/ListViewItemContents.js]: Instead of defaulting to this icon (not sure why I did this in the first place) we should conditionally display the icon container based on whether this was supplied.

**Comment by priley86** [comment]
hey @waldenraines - thanks again for submitting this PR. My vote is to call this component  `ListGroup` instead of `ListView`. I think it is OK to support both for the time being. The `list-group` css is older and now based on the existing PF list view:
https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-rows.html

The new `ListView` in PF Core has different CSS classes which support flex. It can be found here:
https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-pf.html

These are very different components from a CSS standpoint. 

Also re: `npm run storybook:deploy`. I think it will be easy enough to modify this package to support a target branch. We are also working to modify the Travis build to allow building branch specific outputs (so that you may have multiple PRs open w/ different storybooks). I will try to provide an update on this in the next PF-R Community Meeting.

**Comment by waldenraines** [comment]
> hey @waldenraines - thanks again for submitting this PR. My vote is to call this component ListGroup instead of ListView. I think it is OK to support both for the time being. The list-group css is older and now based on the existing PF list view:
> https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-rows.html
> 
> The new ListView in PF Core has different CSS classes which support flex. It can be found here:
> https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-pf.html
> 
> These are very different components from a CSS standpoint.

Are you saying that what is called [ListView](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view.html), [List View - Simple Expansion](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-simple-expansion.html), and [List View - Compound Expansion](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-compound-expansion.html) on the rawgit tests [here](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/) shouldn't be called list view in patternfly-react?

Wouldn't that necessitate renaming list-view implementations across all projects as well as in PF core?

**Comment by priley86** [comment]
@waldenraines it's a very good point. The problem is we now have conflicting names in our PF Core test pages.

@bleathem @andresgalante we'll need some more advice on how to proceed and name these components appropriately.

**Comment by priley86** [comment]
@waldenraines let's keep the name as you have it here (`ListView`). I will rename my component `List` to match the PF Core test pages for now.

Eventually, ListView will become deprecated and we can convert to List, but this is a ways out...

**Comment by waldenraines** [comment]
> @waldenraines let's keep the name as you have it here (ListView). I will rename my component List to match the PF Core test pages for now.
> 
> Eventually, ListView will become deprecated and we can convert to List, but this is a ways out...

Sounds good to me, thanks.

**Comment by jgiardino** [comment]
Hi @waldenraines,

Thanks for contributing this pattern! I reviewed the implementation of the html and css relative to what’s implemented for PF core css. There were a few issues I noticed and some questions I have:

- In the storybook example [List View - Simple Expansion](https://waldenraines.github.io/patternfly-react/?knob-Label=Danger%20Will%20Robinson%21&selectedKind=ListView&selectedStory=List%20View%20-%20Simple%20Expansion&full=0&down=1&left=1&panelRight=0&downPanel=kadirahq%2Fstorybook-addon-knobs), clicking the Action button triggers the expansion. 
  - Additionally, if a link is included anywhere, I would expect clicking the link to not toggle the expand/collapse state of the list item.

- Clicking in the expanded section of a list item triggers the toggle to the collapsed state. This isn’t consistent with implementations on the PF test pages. In the [PF core example](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-simple-expansion.html), clicking the X in the expanded section is the only area that toggles the state back to collapsed.

- I don’t see an example showing an action button with a kebab. Can you confirm whether the component supports the kebab (e.g. can I click the kebab to show a menu)? 

- Should the react components for buttons and kebabs be used if they are included in the list view?

- I only see checkboxes in the simple expansion view for the 3rd and 6th list items, but the same mock list items are used for each list view example. The non-expandable list item should also support the ability to select a list item.

- Does this implementation allow for other icons besides font awesome (e.g. [PF icons](http://www.patternfly.org/styles/icons/)) or for using other icon classes (e.g. `list-view-pf-icon-lg`, `list-view-pf-icon-md`, `list-view-pf-icon-success`, etc…). Additionally, this [test page](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-rows.html) includes an example where a calendar date displays instead of an icon. Should `list-view-pf-left` be more generic in terms of what contents can be defined there, to support all the examples shown on this page?

**Comment by jgiardino** [comment]
Oooops, didn't mean to hit that close button.

File: src/ListView/ListViewExpandableItem.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            heading={this.props.heading}
+            itemText={this.props.itemText}
+            additionalInfo={this.props.additionalInfo}
+          />
+        </div>
+        <div className={listItemContainerClass}>
+          <div className="close" onClick={this.toggleExpanded}>
+            <span className="pficon pficon-close" />
+          </div>
+          {this.props.expansion}
```
Comment (jtomasek) [inline src/ListView/ListViewExpandableItem.js]: {this.props.expanded && this.props.expansion} This may be needed as I am not sure if hiding the component via css prevents it from rendering. In any case, we want the expansion not to render until the component is expanded.

File: src/ListView/ListViewItem.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import React from 'react'
```
Comment (jtomasek) [inline src/ListView/ListViewItem.js]: Most of the comments I made to ListVIewItemExpanded fits this component too.

File: package.json [hunk 5]
```diff
@@ -48,22 +49,24 @@
...
+    "react": "^15.5.0"
   },
   "scripts": {
     "commit": "git-cz",
     "build": "babel src --out-dir lib --ignore test.js",
     "lint": "eslint --max-warnings 0 src storybook",
     "prettier": "prettier --write --single-quote --no-semi '{src,storybook}/**/*.js'",
     "prepublish": "npm run build",
-    "test": "npm run lint && jest",
+    "test": "rm -rf ./lib && npm run lint && jest",
```
Comment (jtomasek) [inline package.json]: Do you hit any issue without this? I am not able to reproduce any problem without.

**Comment by waldenraines** [comment]
Thanks for the reviews @jgiardino and @jtomasek.  I will work on getting these items addressed next week.

**Comment by priley86** [comment]
@waldenraines i've merged the Storybook3 PR just now... please do rebase here and let me know if you have any troubles ;)

Thanks a lot for the help w/ this!

**Comment by waldenraines** [comment]
Closing in favor of https://github.com/patternfly/patternfly-react/pull/50.
